src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,461 @@
+  std::vector<std::vector<char>>& Description() {return boardDescription;}
+
+ private:
+  /**
+   * Utility function that helps generate random environment board.
+   * 
+   * @return board, a 2D array of characters that hold the board description.
+   *    'S' is the Start Tile, the starting position of the agent.
+   *    'G' is the Goal Tile, the agent can walk into a goal tile to win.
+   *    'F' is a Frozen Tile, the agent can walk on it.
+   *    'H' is a Hole Tile, the agent can fall into it and the game ends.
+   */
+  std::vector<std::vector<char>> generateRandomBoard()
+  {
+    bool valid = false;
+    std::vector<std::vector<char>> Board;
+    //! TODO: Tentative, implement max step to control whether
+    // we can never generate a possible board.
+    while (!valid)
+    {
+      // Randomly choose tiles in the board (discrete random distribution).
+      Board = genBoardHelper(height, width, platformRate);
+      valid = dfsHelper(Board, height, width);
+    }
+    return Board;
+  }
+
+  /**
+   * Perform depth-firsth search to see if the environment board has 
+   * a solution or not.
+   * 
+   * @param candidateBoard Environment board description.
+   * @param height Height of the environment board.
+   * @param width Width of the environment board.
+   * @return true if there is a solution, false otherwise. 
+   */
+  static bool dfsHelper(std::vector<std::vector<char>> candidateBoard, size_t height, size_t width)
+  {
+    arma::Mat<short> visited(height, width);
+    visited.fill(0);
+    std::vector<std::array<size_t, 2>> path;
+    path.push_back({0, 0});
+    while (!path.empty())
+    {
+      auto node = path.back();
+      path.pop_back();
+
+      visited(node[0],node[1]) = 1;
+      int directions[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
+      for (auto direction : directions)
+      {
+        size_t r_new = direction[0] + node[0];
+        size_t c_new = direction[1] + node[1];
+        if (!visited(r_new, c_new))
+        {
+          if (r_new >= height || c_new >= width)
+              continue;
+          if (candidateBoard[r_new][c_new] == 'G')
+              return true;
+          if (candidateBoard[r_new][c_new] != 'H')
+              path.push_back({r_new, c_new});
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Perform random distribution of tile in the environment board. There are 4 type of tile:
+   * 'S' is the Start Tile, the starting position of the agent.
+   * 'G' is the Goal Tile, the agent can walk into a goal tile to win.
+   * 'F' is a Frozen Tile, the agent can walk on it.
+   * 'H' is a Hole Tile, the agent can fall into it and the game ends.
+   * 
+   * @param m Height of the environment board.
+   * @param n Width of the environment board.
+   * @return a 2d array that describes the environment board. 
+   */
+  std::vector<std::vector<char>> genBoardHelper(size_t height, size_t width, double platformRate)
+  {
+    std::vector<char> board[height];
```

Comment:
I am just curious to know why the board is represented in `std::vector`? is there any advantages over `armadillo`?
If the dimension of the board is always small, it might not be an issue, but if the dimensions increase, I would see some performance issues.
Let me know what do you think

Reply:
Agreed, there is no benefit from using `std::vector` I would even say using `arma::Mat` is more intuitive, since you don't need a vector for height and width.

Reply:
I see, but since we cannot use arma::Mat for character type, maybe I will refactor the function to use unsigned integer instead.

Reply:
Yeah, unsigned int should be just fine.

Reply:
I have added an enum there and use arma::Mat<size_t> to represent the board. The test passed.